x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/AuthenticationService.java

```diff
@@ -208,18 +267,21 @@ long getNumInvalidation() {
         private AuthenticationToken authenticationToken = null;
         private AuthenticationResult authenticationResult = null;
 
-        Authenticator(RestRequest request, ActionListener<Authentication> listener) {
-            this(new AuditableRestRequest(auditTrail, failureHandler, threadContext, request), null, listener);
+        Authenticator(RestRequest request, boolean fallbackToAnonymous, ActionListener<Authentication> listener) {
+            this(new AuditableRestRequest(auditTrail, failureHandler, threadContext, request), null, fallbackToAnonymous, listener);
         }
 
-        Authenticator(String action, TransportMessage message, User fallbackUser, ActionListener<Authentication> listener) {
-            this(new AuditableTransportRequest(auditTrail, failureHandler, threadContext, action, message
-            ), fallbackUser, listener);
+        Authenticator(String action, TransportMessage message, User fallbackUser, boolean fallbackToAnonymous,
```

Comment:
Since `fallbackToAnonymous` makes no sense when `fallbackUser` is passed in and not null, I'm wondering if this is clearer to have a ctor and a `#createAuthenticator()` without the  `fallbackToAnonymous` parameter and Objects.notNull for the `fallbackUser`. It feels like it would be simpler to argue about or understand while reading the code.

Reply:
The other option is to merge the two options (`fallbackUser` and `fallbacktoAnonymous`) into a single field in some way. Right now those two cases in `handleNullToken` are almost identical. 

There's just enough minor differences to make it a bit tricky though.
I could make it a `Supplier<Authentication> fallback`, but I don't think I'd get anything cleaner than that.

Reply:
@jkakavas Let me know if you'd like to see that change. I've implemented it as you suggested for now, but could do more.

Reply:
I think this is clean enough. Maybe we can reconsider additional changes in the refactoring effort that will follow at some point